Prisma throws type mismatch error using Decimal type from from 'decimal.js' package

## Bug description

Prisma generates a namespace for Decimal.js that does not match the type from the import of Decimal.js from npm. This makes it annoying to use when third party code gives you "decimal.js Decimal's" and not "prisma Decimal's".

```
Type 'Decimal[]' is not assignable to type 'string | number | Decimal | Decimal[] | TestCreatefooInput | number[] | string[]'
```

## How to reproduce

```

model Test {
  id  Int       @id @default(autoincrement())
  foo Decimal[]
}
```

```
iimport { PrismaClient } from '@prisma/client'

import {Decimal as PrismaDecimal} from '@prisma/client/runtime'
import {Decimal as DecimalActual} from 'Decimal.js'

const prisma = new PrismaClient()

const actualDecimals = [new DecimalActual(1), new DecimalActual(2), new DecimalActual(3)]
const prismaDecimals = [new PrismaDecimal(1), new PrismaDecimal(2), new PrismaDecimal(3)]

prisma.test.create({
	data: {
		foo: actualDecimals
	  // ^ Type 'Decimal[]' is not assignable to type 'string | number | Decimal | Decimal[] | TestCreatefooInput | number[] | string[]'
	}
})

prisma.test.create({
	data: {
		foo: prismaDecimals
	}
})

```

## Expected behavior

Doesn't complain about mismatched types

## Prisma information

{
  "@prisma/client": "^2.18.0",
  "decimal.js": "^10.2.1"
}

## Environment & setup

```
Environment variables loaded from .env
prisma               : 2.18.0
@prisma/client       : 2.18.0
Current platform     : darwin
Query Engine         : query-engine da6fafb57b24e0b61ca20960c64e2d41f9e8cff1 (at ../../.config/yarn/global/node_modules/@prisma/engines/query-engine-darwin)
Migration Engine     : migration-engine-cli da6fafb57b24e0b61ca20960c64e2d41f9e8cff1 (at ../../.config/yarn/global/node_modules/@prisma/engines/migration-engine-darwin)
Introspection Engine : introspection-core da6fafb57b24e0b61ca20960c64e2d41f9e8cff1 (at ../../.config/yarn/global/node_modules/@prisma/engines/introspection-engine-darwin)
Format Binary        : prisma-fmt da6fafb57b24e0b61ca20960c64e2d41f9e8cff1 (at ../../.config/yarn/global/node_modules/@prisma/engines/prisma-fmt-darwin)
Studio               : 0.356.0
```

Preface: As you might note below, I'm using a custom output folder for my client, not generating it into @prisma/client.

Because these two types are not the same, I attempted to use Prisma.Decimal on my angular frontend to maintain strong typing.  However when I do that and attempt to server my angular application I'm met with a slew of message similar to
```
Error: ./libs/stats-models/.prisma/client/runtime/index.js
Module not found: Error: Can't resolve 'zlib' in 'D:\Monorepo\libs\stats-models\.prisma\client\runtime'
```

--- a/packages/client/src/__tests__/decimalJsLike.test.ts
+++ b/packages/client/src/__tests__/decimalJsLike.test.ts
@@ -1,0 +1,85 @@
+import Decimal from 'decimal.js'
+
+import { isDecimalJsLike, stringifyDecimalJsLike } from '../runtime/utils/decimalJsLike'
+
+describe('isDecimalJsLike', () => {
+  test('true for decimal.js instance', () => {
+    expect(isDecimalJsLike(new Decimal('12.3'))).toBe(true)
+  })
+
+  test('false for null', () => {
+    expect(isDecimalJsLike(null)).toBe(false)
+  })
+
+  test('false for primitives', () => {
+    expect(isDecimalJsLike(1)).toBe(false)
+    expect(isDecimalJsLike('one')).toBe(false)
+    expect(isDecimalJsLike(true)).toBe(false)
+    expect(isDecimalJsLike(Symbol())).toBe(false)
+    expect(isDecimalJsLike(BigInt('123'))).toBe(false)
+  })
+
+  test('true for decimal.js-like object', () => {
+    const object = {
+      d: [12, 3000000],
+      e: 1,
+      s: 1,
+    }
+    expect(isDecimalJsLike(object)).toBe(true)
+  })
+
+  test('false for object with incorrect `d` property', () => {
+    const object = {
+      d: 'yes',
+      e: 1,
+      s: 1,
+    }
+    expect(isDecimalJsLike(object)).toBe(false)
+  })
+
+  test('false for object with incorrect `e` property', () => {
+    const object = {
+      d: [12, 3000000],
+      e: 'one',
+      s: 1,
+    }
+    expect(isDecimalJsLike(object)).toBe(false)
+  })
+
+  test('false for object with incorrect `s` property', () => {
+    const object = {
+      d: [12, 3000000],
+      e: 1,
+      s: '+',
+    }
+    expect(isDecimalJsLike(object)).toBe(false)
+  })
+
+  test('allows to have extra properties', () => {
+    const object = {
+      d: [12, 3000000],
+      e: 1,
+      s: 1,
+      something: 'other',
+      isFinite() {
+        return true
+      },
+    }
+    expect(isDecimalJsLike(object)).toBe(true)
+  })
+})
+
+describe('stringifyDecimalJsLike', () => {
+  test('stringifies Decimal instance', () => {
+    expect(stringifyDecimalJsLike(new Decimal('12.3'))).toBe('12.3')
+  })
+
+  test('stringifies Decimal-like instance', () => {
+    const object = {
+      d: [12, 3000000],
+      e: 1,
+      s: 1,
+    }
+    expect(stringifyDecimalJsLike(object)).toBe('12.3')
+  })
+})

--- a/packages/client/src/__tests__/query/decimal.test.ts
+++ b/packages/client/src/__tests__/query/decimal.test.ts
@@ -40,6 +40,34 @@ test('allows to pass it decimal instance', () => {
   `)
 })
 
+test('allows to pass it decimal-like object', () => {
+  const document = makeDocument({
+    dmmf,
+    rootTypeName: 'query',
+    rootField: 'findManyUser',
+    select: {
+      where: {
+        money: {
+          d: [12, 5000000],
+          e: 1,
+          s: 1,
+        },
+      },
+    },
+  })
+
+  expect(document.toString()).toMatchInlineSnapshot(`
+    query {
+      findManyUser(where: {
+        money: 12.5
+      }) {
+        id
+        money
+      }
+    }
+  `)
+})
+
 test('allows to pass it decimal array', () => {
   const document = makeDocument({
     dmmf,

--- a/packages/client/src/generation/TSClient/common.ts
+++ b/packages/client/src/generation/TSClient/common.ts
@@ -28,7 +28,8 @@ const {
   empty,
   join,
   raw,
-  Decimal
+  Decimal,
+  DecimalJsLike
 } = require('${runtimeDir}/${runtimeName}')
 `
 }
@@ -112,6 +113,8 @@ export import Sql = runtime.Sql
  * Decimal.js
  */
 export import Decimal = runtime.Decimal
+
+export type DecimalJsLike = runtime.DecimalJsLike
 
 /**
  * Prisma Client JS version: ${clientVersion}

--- a/packages/client/src/runtime/index.ts
+++ b/packages/client/src/runtime/index.ts
@@ -5,6 +5,7 @@ export { DMMF } from './dmmf-types'
 export type { PrismaClientOptions } from './getPrismaClient'
 export { getPrismaClient } from './getPrismaClient'
 export { makeDocument, PrismaClientValidationError, transformDocument, unpack } from './query'
+export type { DecimalJsLike } from './utils/decimalJsLike'
 export { findSync } from './utils/find'
 export { warnEnvConflicts } from './warnEnvConflicts'
 export {

--- a/packages/client/src/runtime/query.ts
+++ b/packages/client/src/runtime/query.ts
@@ -25,6 +25,7 @@ import {
   unionBy,
   wrapWithList,
 } from './utils/common'
+import { isDecimalJsLike, stringifyDecimalJsLike } from './utils/decimalJsLike'
 import { deepExtend } from './utils/deep-extend'
 import { deepGet } from './utils/deep-set'
 import { filterObject } from './utils/filterObject'
@@ -635,8 +636,8 @@ function stringify(value: any, inputType?: DMMF.SchemaArgInputType) {
     return 'null'
   }
 
-  if (Decimal.isDecimal(value)) {
-    return value.toString()
+  if (Decimal.isDecimal(value) || (inputType?.type === 'Decimal' && isDecimalJsLike(value))) {
+    return stringifyDecimalJsLike(value)
   }
 
   if (inputType?.location === 'enumTypes' && typeof value === 'string') {

--- a/packages/client/src/runtime/utils/common.ts
+++ b/packages/client/src/runtime/utils/common.ts
@@ -5,6 +5,7 @@ import leven from 'js-levenshtein'
 
 import type { DMMFHelper } from '../dmmf'
 import type { DMMF } from '../dmmf-types'
+import { isDecimalJsLike } from './decimalJsLike'
 
 export interface Dictionary<T> {
   [key: string]: T
@@ -71,7 +72,7 @@ export const GraphQLScalarToJSTypeTable = {
   UUID: 'string',
   Json: 'JsonValue',
   Bytes: 'Buffer',
-  Decimal: ['Decimal', 'number', 'string'],
+  Decimal: ['Decimal', 'DecimalJsLike', 'number', 'string'],
   BigInt: ['bigint', 'number'],
 }
 
@@ -117,6 +118,10 @@ export function getGraphQLType(value: any, potentialType?: string | DMMF.SchemaE
 
   // https://github.com/MikeMcl/decimal.js/blob/master/decimal.js#L4499
   if (Decimal.isDecimal(value)) {
+    return 'Decimal'
+  }
+
+  if (potentialType === 'Decimal' && isDecimalJsLike(value)) {
     return 'Decimal'
   }
 

--- a/packages/client/src/runtime/utils/decimalJsLike.ts
+++ b/packages/client/src/runtime/utils/decimalJsLike.ts
@@ -1,0 +1,36 @@
+import { Decimal } from 'decimal.js'
+/**
+ * Interface for any Decimal.js-like library
+ * Allows us to accept Decimal.js from different
+ * versions and some compatible alternatives
+ */
+export interface DecimalJsLike {
+  d: number[]
+  e: number
+  s: number
+}
+
+export function isDecimalJsLike(value: unknown): value is DecimalJsLike {
+  if (Decimal.isDecimal(value)) {
+    return true
+  }
+  return (
+    value !== null &&
+    typeof value === 'object' &&
+    typeof (value as any).s === 'number' &&
+    typeof (value as any).e === 'number' &&
+    Array.isArray((value as any).d)
+  )
+}
+
+export function stringifyDecimalJsLike(value: DecimalJsLike): string {
+  if (Decimal.isDecimal(value)) {
+    return String(value)
+  }
+
+  const tmpDecimal = new Decimal(0) as DecimalJsLike
+  tmpDecimal.d = value.d
+  tmpDecimal.e = value.e
+  tmpDecimal.s = value.s
+  return String(tmpDecimal)
+}

--- a/packages/client/tests/functional/_utils/idForProvider.ts
+++ b/packages/client/tests/functional/_utils/idForProvider.ts
@@ -1,0 +1,6 @@
+const sqlId = 'Int @id @default(autoincrement())'
+const mongoDbId = 'String @id @default(auto()) @map("_id") @db.ObjectId'
+
+export function idForProvider(provider: string): string {
+  return provider === 'mongodb' ? mongoDbId : sqlId
+}

--- a/packages/client/tests/functional/_utils/idForProvider.ts
+++ b/packages/client/tests/functional/_utils/idForProvider.ts
@@ -1,4 +1,4 @@
-const sqlId = 'Int @id @default(autoincrement())'
+const sqlId = 'String @id @default(uuid())'
 const mongoDbId = 'String @id @default(auto()) @map("_id") @db.ObjectId'
 
 export function idForProvider(provider: string): string {

--- a/packages/client/tests/functional/decimal/_matrix.ts
+++ b/packages/client/tests/functional/decimal/_matrix.ts
@@ -1,0 +1,13 @@
+export default () => [
+  [
+    {
+      provider: 'sqlite',
+    },
+    {
+      provider: 'postgresql',
+    },
+    {
+      provider: 'mysql',
+    },
+  ],
+]

--- a/packages/client/tests/functional/decimal/prisma/_schema.ts
+++ b/packages/client/tests/functional/decimal/prisma/_schema.ts
@@ -1,0 +1,19 @@
+import { idForProvider } from '../../_utils/idForProvider'
+
+export default ({ provider }) => {
+  return /* Prisma */ `
+  generator client {
+    provider = "prisma-client-js"
+  }
+  
+  datasource db {
+    provider = "${provider}"
+    url      = env("DATABASE_URI_${provider}")
+  }
+  
+  model User {
+    id ${idForProvider(provider)}
+    money Decimal
+  }
+  `
+}

--- a/packages/client/tests/functional/decimal/tests.ts
+++ b/packages/client/tests/functional/decimal/tests.ts
@@ -1,0 +1,58 @@
+import { Decimal } from 'decimal.js'
+
+import { setupTestSuiteMatrix } from '../_utils/setupTestSuiteMatrix'
+
+// @ts-ignore this is just for type checks
+declare let prisma: import('@prisma/client').PrismaClient
+
+setupTestSuiteMatrix(() => {
+  beforeAll(async () => {
+    await prisma.user.create({
+      data: { money: new Decimal('12.5') },
+    })
+  })
+
+  test('decimal as Decimal.js instance', async () => {
+    const result = await prisma.user.findFirst({
+      where: {
+        money: new Decimal('12.5'),
+      },
+    })
+
+    expect(String(result?.money)).toBe('12.5')
+  })
+
+  test('decimal as string', async () => {
+    const result = await prisma.user.findFirst({
+      where: {
+        money: '12.5',
+      },
+    })
+
+    expect(String(result?.money)).toBe('12.5')
+  })
+
+  test('decimal as number', async () => {
+    const result = await prisma.user.findFirst({
+      where: {
+        money: { gt: 12.4, lt: 12.6 },
+      },
+    })
+
+    expect(String(result?.money)).toBe('12.5')
+  })
+
+  test('decimal as decimal.js-like object', async () => {
+    const result = await prisma.user.findFirst({
+      where: {
+        money: {
+          d: [12, 5000000],
+          e: 1,
+          s: 1,
+        },
+      },
+    })
+
+    expect(String(result?.money)).toBe('12.5')
+  })
+})

